src/models/ModelBuilder.cc

```diff
@@ -3,23 +3,136 @@
 //
 
 #include "ModelBuilder.h"
+#include <unordered_set>
 
-void ModelBuilder::processSubstitutions(Logic::SubstMap const & subst) {
-    Logic::SubstMap copy;
-    subst.copyTo(copy);
-    logic.substitutionsTransitiveClosure(copy);
-    auto assignCopy = this->assignment;
-    auto model = this->build();
-    for (auto key : copy.getKeys()) {
-        if (logic.isIte(key)) {
-            // We store only values of reals variables
+std::unique_ptr<Model> ModelBuilder::build() {
+    std::unordered_map<SymRef,TemplateFunction,SymRefHash> builtDefinitions;
+    for (auto & symbolSigVal : definitions) {
+        const ValuationNode * valuationNode = symbolSigVal.second.second;
+        SRef sr = logic.getSortRef(symbolSigVal.first);
+        PTRef body = valuationTreeToFunctionBody(valuationNode, sr);
+        TemplateFunction templateFun(std::move(symbolSigVal.second.first), body);
+        builtDefinitions[symbolSigVal.first] = std::move(templateFun);
+    }
+    return std::unique_ptr<Model>(new Model(logic, assignment, builtDefinitions));
+}
+
+void ModelBuilder::addToTheoryFunction(SymRef sr, vec<PTRef> vals, PTRef val)
+{
+    if (logic.getSortRef(sr) != logic.getSortRef(val)) {
+        throw OsmtApiException("Incompatible sort for symbol valuation");
+    }
+    Symbol & sym = logic.getSym(sr);
+    if (sym.nargs() != vals.size_()) {
+        throw OsmtApiException("Incorrect valuation for symbol: argument and valuation size do not match");
+    }
+    for (int i = 0; i < vals.size(); i++) {
+        if (sym[i] != logic.getSortRef(vals[i])) {
+            throw OsmtApiException("Incorrect valuation for symbol: sort mismatch");
+        }
+    }
+
+    if (not hasTheoryFunction(sr)) {
+        vec<PTRef> formalArgs; formalArgs.capacity(vals.size());
+        std::string symName(logic.getSymName(sr));
+        // Ensure that no formal arg name collides with the function name
+        std::string formalArgPrefix(Model::getFormalArgBaseNameForSymbol(logic, sr, formalArgDefaultPrefix));
+
+        for (PTRef v : vals) {
+            std::stringstream ss;
+            ss << formalArgPrefix << uniqueNum++;
+            formalArgs.push(logic.mkVar(logic.getSortRef(v), ss.str().c_str()));
+        }
+        FunctionSignature templateSig(logic.protectName(logic.getSymName(sr)), std::move(formalArgs), logic.getSortRef(sr));
+        definitions.insert({sr,opensmt::pair<FunctionSignature,ValuationNode*>{std::move(templateSig), nullptr}});
+    }
+    auto & signatureAndValuation = definitions.at(sr);
+    FunctionSignature & templateSignature = signatureAndValuation.first;
```

Comment:
`templateSignature` can be inlined, it has only one usage.

Reply:
Ok.